File: robot_controller/robot_controller/PidController.py
import time

class PidController():
    def __init__(self, Kp, Ki, Kd):
        #PID params
        self.Kp = Kp
        self.Ki = Ki
        self.Kd = Kd
        
        #PID values
        self.previous_error = 0
        self.integral = 0
        self.last_time = time.time()


    def compute(self, error, current_time):
        # PID terms
        P = self.Kp * error

        self.integral += error
        I = self.Ki * self.integral

        dedt = (error - self.previous_error) / (current_time - self.last_time)
        D = self.Kd * dedt
        self.last_time = current_time
        self.previous_error = error

        output = P + I + D

        return output

File: robot_controller_cpp/src/PidController.cpp
#include "PidController.hpp"
#include <chrono>

PidController::PidController(double Kp, double Ki, double Kd)
    : Kp_(Kp), Ki_(Ki), Kd_(Kd), previous_error_(0.0), integral_(0.0)
{
    last_time_ = std::chrono::steady_clock::now(); // Initialize time
}

double PidController::compute(double error)
{
    // Get current time
    auto current_time = std::chrono::steady_clock::now();
    std::chrono::duration<double> elapsed_time = current_time - last_time_;
    double dt = elapsed_time.count(); // Duration in seconds

    // Prevent division by zero
    if (dt <= 0.0)
    {
        dt = 1e-6; // Small value
    }

    // Proportional term
    double P = Kp_ * error;

    // Integral term
    integral_ += error * dt;
    double I = Ki_ * integral_;

    // Derivative term
    double derivative = (error - previous_error_) / dt;
    double D = Kd_ * derivative;

    // Save error and time for next iteration
    previous_error_ = error;
    last_time_ = current_time;

    // PID output
    return P + I + D;
}
